pkg/stewardlabels/propagate_test.go

```diff
@@ -0,0 +1,508 @@
+				nil,
+				{},
+				{
+					other1: other1Val,
+				},
+				{
+					enfor1: enfor1Val,
+					other1: other1Val,
+				},
+				{
+					enfor1: enfor1Val,
+					enfor2: enfor2Val,
+					other1: other1Val,
+				},
+			} {
+				testcases = append(testcases, PropagateSuccessTestCase{
+					PreexistingDestLabels: destLabels,
+					SourceLabels:          sourceLabels,
+					LabelSpec:             labelSpec,
+					ExpectedResultLabels:  expectedDestLabels,
+				})
+			}
+		}
+
+		// execute
+		testcases.run(t, "WithAdditionalLabelInDestObj/")
+	}
+}
+
+type PropagateFailsTestCase struct {
+	// fields must be exported to make deepcopy work
+
+	DestLabels   map[string]string
+	SourceLabels map[string]string
+	LabelSpec    map[string]string
+
+	ExpectedErrorMsg string
+}
+
+type PropagateFailsTestCases []PropagateFailsTestCase
+
+func (testcases PropagateFailsTestCases) run(t *testing.T, namePrefix string) {
+	t.Helper()
+
+	for idx, tc := range testcases {
+		t.Run(fmt.Sprintf("%s%d", namePrefix, idx), func(t *testing.T) {
+			t.Helper()
+
+			tc := deepcopy.Copy(tc).(PropagateFailsTestCase)
+			t.Parallel()
+			logTestCaseSpec(t, tc)
+
+			// SETUP
+			destLabels := deepcopy.Copy(tc.DestLabels).(map[string]string)
+			destObj := &DummyObject1{}
+			destObj.SetLabels(destLabels)
+
+			sourceObj := &DummyObject1{}
+			sourceObj.SetLabels(tc.SourceLabels)
+
+			// EXERCISE
+			resultErr := propagate(destObj, sourceObj, tc.LabelSpec)
+
+			// VERIFY
+			assert.Error(t, resultErr, tc.ExpectedErrorMsg)
+			assert.Assert(t, cmp.DeepEqual(tc.DestLabels, destObj.GetLabels()),
+				"labels have been modified")
+			if destLabels != nil {
+				assert.Assert(t, testableLabelMap(destObj.GetLabels()).IsSameMapAs(destLabels),
+					"label map of destination object has been replaced")
+			}
+		})
+	}
+}
+
+func expectedErrorMessageForValueConflictAtDestObj(labelKey, existingValue, expectedValue string) string {
+	return fmt.Sprintf(
+		"value conflict: destination object label %q has existing value %q but %q is expected",
+		labelKey, existingValue, expectedValue,
+	)
+}
+
+func expectedErrorMessageForValueConflictAtSourceObj(labelKey, existingValue, expectedValue string) string {
+	return fmt.Sprintf(
+		"value conflict: source object label %q has value %q but %q is expected",
+		labelKey, existingValue, expectedValue,
+	)
+}
+
+func Test__propagate__Fails_ValueConflictOnDestObj_FromPropagation(t *testing.T) {
+	const (
+		conflict1    = "conflict1Key"
+		conflict1Val = "conflict1Val"
+	)
+
+	testcases := PropagateFailsTestCases{}
+
+	testcases = append(testcases, PropagateFailsTestCase{
+		DestLabels: map[string]string{
+			conflict1: conflict1Val + "-dest",
+		},
+		SourceLabels: map[string]string{
+			conflict1: conflict1Val + "-source",
+		},
+		LabelSpec: map[string]string{
+			conflict1: "",
+		},
+
+		ExpectedErrorMsg: expectedErrorMessageForValueConflictAtDestObj(
+			conflict1, conflict1Val+"-dest", conflict1Val+"-source",
+		),
+	})
+
+	testcases.run(t, "")
```

Comment:
Name prefix is always empty string. Is this intended? Then you don't need it at all.

Reply:
I applied the same pattern as for the other `(...TestCases) run()` methods. For that method each set of test cases is in a separate top-level test function, so there's no need for an additional subtest name prefix.

I've removed the unused parameter in bcb69f4.